src/Elasticsearch.Net/Transport/PostData.cs

```diff
@@ -87,91 +146,159 @@ protected internal PostData(IEnumerable<object> item)
 		{
-			MemoryStream ms = null;
+			MemoryStream buffer = null;
+			var stream = writableStream;
+			var disableDirectStreaming = DisableDirectStreaming ?? settings.DisableDirectStreaming;
+
 			switch (Type)
 			{
 				case PostType.ByteArray:
-					ms = settings.MemoryStreamFactory.Create(WrittenBytes);
+					if (!disableDirectStreaming)
+						await stream.WriteAsync(WrittenBytes, 0, WrittenBytes.Length, cancellationToken).ConfigureAwait(false);
+					else
+						buffer = settings.MemoryStreamFactory.Create(WrittenBytes);
+					break;
+
+#if NETSTANDARD2_1
+				case PostType.ReadOnlyMemory:
+					if (_memoryOfBytes.IsEmpty) return;
+
+					if (!disableDirectStreaming)
+						stream.Write(_memoryOfBytes.Span);
+					else
+					{
+						WrittenBytes ??= _memoryOfBytes.Span.ToArray();
+						buffer = settings.MemoryStreamFactory.Create(WrittenBytes);
+					}
 					break;
+#endif
+
 				case PostType.LiteralString:
-					ms = !string.IsNullOrEmpty(_literalString) ? settings.MemoryStreamFactory.Create(_literalString.Utf8Bytes()) : null;
+					if (string.IsNullOrEmpty(_literalString)) return;
+
+					var stringBytes = WrittenBytes ?? _literalString.Utf8Bytes();
+					WrittenBytes ??= stringBytes;
+					if (!disableDirectStreaming)
+						await stream.WriteAsync(stringBytes, 0, stringBytes.Length, cancellationToken).ConfigureAwait(false);
+					else
+						buffer = settings.MemoryStreamFactory.Create(stringBytes);
 					break;
+
 				case PostType.EnumerableOfString:
-					ms = _enumerableOfStrings.HasAny()
-						? settings.MemoryStreamFactory.Create((string.Join(NewLineString, _enumerableOfStrings) + NewLineString).Utf8Bytes())
-						: null;
+					if (!_enumerableOfStrings.HasAny()) return;
+
+					BufferIfNeeded(settings, ref buffer, ref stream);
+					foreach (var s in _enumerableOfStrings)
+					{
+						var bytes = s.Utf8Bytes();
+						await stream.WriteAsync(bytes, 0, bytes.Length, cancellationToken).ConfigureAwait(false);
+						await stream.WriteAsync(NewLineByteArray, 0, 1, cancellationToken).ConfigureAwait(false);
+					}
 					break;
+
 				case PostType.EnumerableOfObject:
 					if (!_enumerableOfObject.HasAny()) return;
 
-					Stream stream;
-					if (DisableDirectStreaming ?? settings.DisableDirectStreaming)
-					{
-						ms = settings.MemoryStreamFactory.Create();
-						stream = ms;
-					}
-					else stream = writableStream;
+					BufferIfNeeded(settings, ref buffer, ref stream);
 					foreach (var o in _enumerableOfObject)
 					{
 						await settings.RequestResponseSerializer.SerializeAsync(o, stream, SerializationFormatting.None, cancellationToken)
 							.ConfigureAwait(false);
 						await stream.WriteAsync(NewLineByteArray, 0, 1, cancellationToken).ConfigureAwait(false);
 					}
 					break;
+
+				case PostType.StreamHandler:
+					throw new Exception("PostData is not expected/capable to handle streamable data, use StreamableData instead");

```

Comment:
I think the exception message should be changed in a similar manner to the comment above.